### PR TITLE
[Scalarizer][DirectX] support structs return types

### LINKED ISSUE #111437: [DirectX][Scalarizer] Add support to scalarize struct of vector return types
Implement this RFC https://discourse.llvm.org/t/rfc-allow-the-scalarizer-pass-to-scalarize-vectors-returned-in-structs/82306

look into complexity of adding `llvm.frexp` test case.

### PR DESCRIPTION
**Pull request by farzonl**
Based on this RFC: https://discourse.llvm.org/t/rfc-allow-the-scalarizer-pass-to-scalarize-vectors-returned-in-structs/82306

LLVM intrinsics do not support out params. To get around this limitation implementers will make intrinsics return structs to capture a return type and an out param. This implementation detail should not impact scalarization since these cases should be elementwise operations.

## Three changes are needed. 
- The CallInst visitor needs to be updated to handle Structs
- A new visitor is needed for `ExtractValue` instructions
- finsh needs to be update to handle structs so that insert elements are properly propogated.

## Testing changes
- Add support for `llvm.frexp`
- Add support for `llvm.dx.splitdouble`

fixes https://github.com/llvm/llvm-project/issues/111437

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:690]: Making this vectorizable was out of scope for this work so did not add `frexp` to `isTriviallyVectorizable`. We should do a follow up task for this.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (inbelic) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:733]: nit: Is there a reason for this to be re-defined? I don't see any side-effects in `getVectorSplit/isStructAllVectors`

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:733]: Nice catch, this is left over from a previous version where`CallType`  was `CI.getType()` or `CI.getType()->getContainedType(0);` and for this case I needed the struct to iterate over. Later change I inited VS as `getVectorSplit(CallType);` or `VS(getVectorSplit(CallType->getContainedType(0));` and thus didn't need to do this, but forgot to remove it.

**Comment by farzonl** [comment]
@nikic do you have some time do to a review?

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (bogner) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:1264]: I think assigning null here may suppress a "not all paths initialize this variable" type warning if we were to add a code path that doesn't. Given that all of the code paths still do initialize the variable that might not be ideal.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (bogner) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:1289]: I think something like `Twine(Op->getName()) + ".elem" + I` might work here and be slightly more efficient than creating a `std::string`

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:690]: I'll add https://github.com/llvm/llvm-project/issues/112408 to the TODO on line 688.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:738]: It is safe because we are within the AreAllVectors block. CurrVS is only nullopt when the input is not a vector.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:202]: I'd prefer not to do this change. If there was a property, I needed on StructType sure. But as it stands it just looks like creating a new variable for no reason.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (bogner) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:738]: This might not be true if the ScalarizeMinBits option for the scalarizer pass is true.

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:738]: I’ve never seen the `if (Split.NumPacked >= NumElems)` case be true.  But sure you are right. We could add an assert or return false?

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:1264]: Reflexive change sorry. Uninitialized reads were a pattern I use to look for in security reviews. But looks like we never do anything but assign to Res so erroneous thinking on my part. Wasn't thinking this might be done this way to catch new conditions.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (bogner) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:739]: I think just returning false is probably the right thing to do. Probably a check like the `if (!OpVS || OpVS->NumPacked != VS->NumPacked)` case below that comments on ScalarizeMinBits would be best.

**Review by bogner — APPROVED** [review]

File: llvm/test/Transforms/Scalarizer/frexp.ll
Comment (nikic) [inline llvm/test/Transforms/Scalarizer/frexp.ll:1]: Please generate this using update_test_checks.py.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:1097]:
```suggestion
        Op0[OpIdx], Index, EVI.getName() + ".elem" + Twine(Index));
```

File: llvm/test/Transforms/Scalarizer/frexp.ll
Comment (nikic) [reply llvm/test/Transforms/Scalarizer/frexp.ll:1]: Could you please also add a test where the struct is returned from the function (not using extractelement)?

**Review by nikic — COMMENTED** [review]
Didn't review this in detail, but looks sensible.

File: llvm/test/Transforms/Scalarizer/frexp.ll
Comment (farzonl) [reply llvm/test/Transforms/Scalarizer/frexp.ll]: add a `test_vector_frexp_void`. It only does the calls.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: There seems to be a subtle potential issue.  `VS` is presumed to be the return vector type for the function.  Later, we use `if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))` to determine whether the overload type is based on the return type, and will use this contained type.  This cannot differentiate between the elements of the struct.

Perhaps that's the intended design for this, but if so it should be made explicit, like:
> The overload type may only be dependent on the first member of the returned structure.  In other words, the other members of that returned structure can not contribute to the overload type of the intrinsic.

I think there may be another subtle problem with the component type of the other vector(s) in the returned structure.  What types must they use?  Should they match the overload type, or can they be different?  Perhaps that's a problem for earlier builtin type checking in clang, and not for the code that transforms the intrinsics, but it does seem like an issue that these other outputs are basically not addressable with the current trivially vectorizable scheme.

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: I think you are confusing a few things.  First the overload  type is not based on the return type. The `isVectorIntrinsicWithOverloadTypeAtArg` has a pass through behavior of  `OpdIdx == -1`  All return indices will be true. Second We build up the return in lines 733 to 748.  `lvm.frexp.v2f64.v2i32` becomes two `lvm.frexp.f64.i32`.

Third the behavior of `isVectorIntrinsicWithOverloadTypeAtArg` isn't relevant for struct return types. The scope of this change is not intended to impact any argument overload cases.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:737]: So, this enforces the same component type across all returned vector results, which answers the question in the prior comment: "Should they match the overload type, or can they be different?".

However, I'm not sure it's necessary for them all to have matching component types, is it?  Though if they differ in bit width, this could create a problem when using `ScalarizeMinBits` to prevent scalarization of small types, since this could create a different `NumPacked` for different returned results.

This same issue would apply to input arguments, and now that I look at how that's handled, I see the comment that it won't handle it if `NumPacked` for the op is different than for `VS` (returned result).

I think it would be ok to allow different components, as long as incorporating the same check as is used for the operands instead of expecting an exact match in type.

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:737]: I would prefer to be more conservative. For one I don't think its necessary to relax this because any different component types should have been determined on the vectorized form of the intrinsic. Second, we don't need anything but an exact component match for our use case so why increase our test burden when it isn't necessary?

Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:737]: Wait, I was confused thinking this was comparing equal.  So, `frexp` returns two different vector types.  The prior call to `isVectorIntrinsicWithOverloadTypeAtArg` determines whether the first type is part of the overload, but then we assume any other types in the return that don't match the first one are also part of the overload type.  That doesn't sound right for the general case either.

Since we don't actually ensure that they match, we could have a situation where `NumPacked` differs between returned vector types, which would be a problem.  This was checked for arguments, bailing if they differ, but we aren't checking this for the return values.  I think we need to do that as well.

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:737]: I think you are still confused. `isVectorIntrinsicWithOverloadTypeAtArg ` doesn't check overloads for return types. Go look at that code path return will always be true.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: `isVectorIntrinsicWithOverloadTypeAtArg` defaults to returning true for `-1` because it's the common case for intrinsic overloads to be based on the return type when it matches the other arguments, but the exceptions are informative. Some of these return true for indices -1 and 0, which means the intrinsic has two overload types, one based on the return type, and one based on the first argument.  Another (`is_fpclass`) returns true only for index 0, which means the overload is based on the first argument, not the return type.

This calls `isVectorIntrinsicWithOverloadTypeAtArg` with `-1` to determine whether the first return type in the struct is used in the overload, then (see my reply to the other comment), we just assume any others that don't match the previous type in that struct are also added to the overload.  I don't think this is correct for the general case.  I think `isVectorIntrinsicWithOverloadTypeAtArg` should be augmented to support individually addressing the fields of the returned struct.  One way would be to extend into the negative space for indices: -1, -2, -3 would refer to three fields in a returned struct, in that order.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:737]: Will discuss directly.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (bogner) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: I think that limiting our handling to cases where all of the vector types nested in the struct type of the return type match is pretty reasonable. Adding handling for cases where we have overloads in some subset of the return types seems like it would be quite complicated and I can't really come up with a realistic scenario where it might happen. The fact that the only intrinsic that exists today that has special treatment of the return type at all is `is_fpclass` is probably evidence that this is rare.

I do think a comment and/or an assert if we find that the vector types of the returned struct type mismatch is appropriate, but let's not go overboard and implement a fully generic solution for something that we only have theoretical evidence for being a problem.

Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: > I think that limiting our handling to cases where all of the vector types nested in the struct type of the return type match is pretty reasonable.

As Farzon points out, the `frexp` example has mismatching vector types in the return struct.  My initial assumption about how this was going to work was incorrect, so this thread may have started out with a bit of confusion.

For `frexp`, each of these are involved in the overload type for the intrinsic, so each must be captured.  The way it works in this PR now is that it captures the first field only if `isVectorIntrinsicWithOverloadTypeAtArg` called with `-1` returns true, then it captures each field after that if it differs from the prior field.  This second part sounds basically like a heuristic, not an explicit property describing a specific intrinsic.

That seems fragile, and it would be easy to break if some intrinsic required something different (even if you think that will never come up).  It's certainly possible, though likely rare, that one of the returned vectors does not contribute to the overload type, even if it differs from the other type.  Other possible scenarios couldn't be handled by this as well.

The compromise Farzon came to, after we met earlier, was to add a different query function for struct return argument fields.  At least it makes things explicit, though it means yet another function needs to be plumbed through TTI as well.

File: llvm/test/CodeGen/DirectX/split-double.ll
Comment (tex3d) [inline llvm/test/CodeGen/DirectX/split-double.ll:13]: Other than the fact that we need something to differentiate vector forms of this intrinsic, I don't think the i32 part is a meaningful overload component of this intrinsic.  It could only ever be two i32 return values and one double input per scalar operation.  I suppose this is just a rhetorical comment, since I can't think of a cleaner way to do it for now.

Comment (farzonl) [reply llvm/test/CodeGen/DirectX/split-double.ll:13]: This isn't the right place for this discussion. The team already went through revisions of what this intrinsic should look like on the split double PR https://github.com/llvm/llvm-project/pull/109331. If you want an improvement It be best until after both these PRs merge.

Also overloads aren't really relevant for this intrinsic so lets not complicate the discussion.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: If I'm being honest  intrinsic behaviors determined by `isVectorIntrinsicWithOverloadTypeAtArg`,  `isVectorIntrinsicWithScalarOpAtArg` and `isTriviallyVectorizable` Is kind of awful.  Adding a new one of these to handle struct return types isn't ideal. For one it means every new intrinsic with struct returns has to get added to some obscure function to get properly scalarized and in the future vectorized. Maybe thats better because it makes scalarization and in the future vectorization more explicit? Personally not a fan.

I'm also not thrilled about having to add a new TTI  api. Its a bunch of boiler plate to express something very simple. Especially since the split double case doesn't need it because the returned struct of vectors are all the same type. 

Second, the fragility IMO is overstated. The code as it is captures the pattern that exist in the current set of llvm intrinsics that represent structs by their component parts.

Third i don't agree with this being a heuristic. That implies the behavior is  best guessing at what the next overload type should be when we are being very determinstic here.

Maybe should talk about  this a bit more. If we still want to do the compromise one area to discuss is  how we can do this without a TTI API for now. I'm thinking maybe the `isStructOfMatchingFixedVectors` api could also check element type and if they are all the same we  just use the existing code path of:
```cpp
 if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))
    Tys.push_back(VS->SplitTy);
```
Which is the path splitdouble is currently taking.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:744]: Skipping failure cases for `CurrVS` should happen outside of the if.  Before, this was under a branch checking if the type differed from the last field, so it was only necessary to check this when it changed, but now, we are gating this on whether this field type contributes to the overload type, so the check needs to be moved out of the if.
Also: we still don't bail if `CurrVS->NumPacked != VS->NumPacked`.

```suggestion
      std::optional<VectorSplit> CurrVS = getVectorSplit(
          cast<FixedVectorType>(CallType->getContainedType(I)));
      // When using ScalarizeMinBits, it is possible for
      // VectorSplit.NumPacked >= NumElems. If that happens a VectorSplit
      // is not returned and we will bailout of handling this call.
      // We also bailout if NumPacked does not match.
      if (!CurrVS || CurrVS->NumPacked != VS->NumPacked)
        return false;
      if (isVectorIntrinsicWithStructReturnOverloadAtField(ID, I))
        Tys.push_back(CurrVS->SplitTy);
```

Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:744]: `CurrVS->NumPacked != VS->NumPacked`. doesn't seem right. Because  frexp first out param can be half, float, or double. the second out param is  an int 32. 

`NumPacked = ScalarizeMinBits / ElemTy->getScalarSizeInBits()`

By default, we won't notice this break because `ScalarizeMinBits ` is an option parameter set to 0. and its integer division. but if the option `ScalarizeMinBits`  is set high enough we will notice a difference in behavior.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:706]: I too have a distaste for this particular property lookup pattern, as I've expressed before.  I think these should be properties defined in the intrinsic tablegen file.  Perhaps an intrinsic (vectorization) property struct/object could be defined from that and retrieved (through TTI as necessary), eliminating the need for plumbing attribute properties one by one.

> Third i don't agree with this being a heuristic.

The code was guessing whether a type is added to the overload types based on whether it was different from the previous type.  That's the heuristic.  If the types matched, that doesn't mean that the type is not used in the overload (in the general sense).

**Review by tex3d — APPROVED** [review]
Looks good.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp
Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp]: Oops, I resolved too quickly, as I thought you took the suggestion.

I still think `|| CurrVS->NumPacked != VS->NumPacked` is correct, for the same reason it's checked here:
https://github.com/llvm/llvm-project/blob/709abacdc350d63c61888607edb28ce272daa0a0/llvm/lib/Transforms/Scalar/Scalarizer.cpp#L710-L718

Comment (tex3d) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp]: Note: I had also modified the comment from `This case does not seem to happen, ...`, since it can happen when `ScalarizeMinBits` is used.

**Review by tex3d — COMMENTED** [review]
Sorry, I approved after thinking you had accepted my suggested change, so I'd say that approval should be contingent on resolving the last comment.

**Review by tex3d — APPROVED** [review]
Thanks for the update!

**Comment by mariusz-sikora-at-amd** [comment]
Hi @farzonl I'm having problems when compiling:

```
define <3 x i32> @test_(<3 x i32> %a, <3 x i32> %b) {
  %r = call { <3 x i32>, <3 x i1> } @llvm.uadd.with.overflow.v3i32(<3 x i32> %b, <3 x i32> %b)
  %el = extractvalue { <3 x i32>, <3 x i1> } %r, 0
  ret <3 x i32> %el
}
```

```
./build/bin/opt -passes='function(scalarizer)' ./build/test_fail.ll -S
opt: ./llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From *) [To = llvm::VectorType, From = llvm::Type]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

Are you planning to add `Intrinsic::uadd_with_overflow` to `ScalarizerVisitor::isTriviallyScalarizable` ?

**Comment by farzonl** [comment]
@mariusz-sikora-at-amd  I wasn't aware of `uadd.with.overflow` when I added this feature. I feel like it should just be adding that intrinsic to `isTriviallyScalarizable ` and `isVectorIntrinsicWithStructReturnOverloadAtField`.

I'm a little suprised you are getting a casting error though. The only non dyn_casts are on `llvm/lib/Transforms/Scalar/Scalarizer.cpp:736` and `1087` and both check the vectors before that cast.

**Comment by mariusz-sikora-at-amd** [comment]
> I'm a little suprised you are getting a casting error though. The only non dyn_casts are on `llvm/lib/Transforms/Scalar/Scalarizer.cpp:736` and `1087` and both check the vectors before that cast.

This is rather assert while we are doing `Scatterer::operator[]` and we are creating `CreateExtractElement` but Type is not a Vector but a Structure. Adding `uadd.with.overflow` to `isTriviallyScalarizable` is 'fixing' this. I think `isVectorIntrinsicWith...` is not needed to be extend because it will produce `uadd.with.overflow.i32.i1` and it will fail on other thing (but this may be only AMDGPU specific?).

**Comment by farzonl** [comment]
> > I'm a little suprised you are getting a casting error though. The only non dyn_casts are on `llvm/lib/Transforms/Scalar/Scalarizer.cpp:736` and `1087` and both check the vectors before that cast.
> 
> This is rather assert while we are doing `Scatterer::operator[]` and we are creating `CreateExtractElement` but Type is not a Vector but a Structure. Adding `uadd.with.overflow` to `isTriviallyScalarizable` is 'fixing' this. 

Thats a bug.  Adding to `isTriviallyScalarizable` is not the right fix. Whats going on is because `uadd.with.overflow` isn't in `isTriviallyScalarizable` we aren't splitting the calls so when we call `Scatterer Op0` it isn't operating on a vector. A correct fix wouldn't  replace extract element unless we scalarized the call. How long can you wait for a fix? I'm at a conference right now so limited with how quickly I can get a fix out.

**Comment by farzonl** [comment]
@mariusz-sikora-at-amd  this bug will be fixed by: https://github.com/llvm/llvm-project/pull/113625.

**Pull request merged**
